server/app/middlewares/websocket_backend.rb

```diff
@@ -75,12 +75,32 @@ def call(env)
     end
   end
 
+  # @param grid [Grid]
+  # @param node_id [String]
+  # @param name [String]
+  # @param attrs [Hash]
+  # @raise [Mongo::Error::OperationFailure] concurrent race
+  # @return [HostNode]
+  def create_node(grid, node_id, name, attrs)
+    node = grid.host_nodes.new(node_id: node_id, name: name, **attrs)
+    node.ensure_unique_name
+
+    if node.name != name
+      logger.warn "rename node #{node_id} on name collision: #{node.name}"
+    end
+
+    # XXX: this can still race with concurrent creates, and result in unique name index conflicts
+    node.save!
```

Comment:
This can still fail with `Mongo::Error::OperationFailure`, but so can the `reserve_node_number`: #2071

The server `WebsocketBackend` does not close the agent websocket connection on unknown errors, so the agent will hang.